active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,272 @@
+
+WinUI:
+
+```cs
+protected override object ProvideValue()
+{
+    return ...
+}
+```
+
+WPF:
+
+```cs
+protected override object ProvideValue(IServiceProvider serviceProvider)
+{
+    return ...
+}
+```
+
+The service provider can be used, for example, to discover the property being assigned to.
+
+This spec is adding a new version of ProvideValue to WinUI that passes a similar service provider. All the APIs match WPF, except where WPF uses the System.IServiceProvider, WinUI will use Microsoft.UI.Xaml.Markup.IXamlServiceProvider.
+
+> To do: Add an Issue to cs/winrt to project IServiceProvider to .Net5 as [System.IServiceProvider](http://msdn.microsoft.com/library/System.IServiceProvider)?
+
+## How markup extensions are used in Xaml markup
+
+A developer defines a custom markup extension by subclassing the [MarkupExtension](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Markup.MarkupExtension) class and overriding the ProvideValue method. This can then be used in Xaml markup: the markup extension can be created using the "{ }" markup expression syntax, and the return value from ProvideValue will be set as the property value.
+
+For example, given this markup extension:
+
+```cs
+public class CurrentDate : MarkupExtension
+{
+    public string Format { get; set; } = "";
+    protected override object ProvideValue()
+    {
+        return DateTime.Now.ToString(Format);
+    }
+}
+```
+
+you can do this in Xaml markup:
+
+```xml
+<TextBlock>
+    Today is:  
+    <Run Text="{local:CurrentDate Format=d}"/>
+</TextBlock>
+```
+
+On an en-US machine, this will produce something like:
+ 
+![MarkupExtension sample](MESample.png)
+
+# Examples
+
+## IProvideValueTarget
+
+The following example shows a custom markup extension whose provided value changes based on the specific property being targeted by the custom markup extension. In this case, the developer wants something that will automatically use an AcrylicBrush for the [Control.Background](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Control.Background) or [Border.Background](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Border.Background) properties, but uses a [SolidColorBrush](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Media.SolidColorBrush) for all other properties. The ProvideValue implementation deterimines what value to return by using the IProvideValueTarget it gets from the IXamlServiceProvider parameter passed to ProvideValue.
+
+```cs
+public class BrushSelectorExtension : MarkupExtension
+{
+    public Color Color { get; set; }
+
+    protected override object ProvideValue(IXamlServiceProvider serviceProvider)
+    {
+        Brush brushToReturn = new SolidColorBrush() { Color = Color }; ;
+
+        var provideValueTarget = serviceProvider.GetService(typeof(IProvideValueTarget))
+            as IProvideValueTarget;

```

Comment:
If you're going to crash if `IProvideValueTarget` is not present, then use a cast instead of `as` so the crash is right at the line of code that caused the problem, instead of deferring the crash until some later point. (Repeat for other uses of "as" where the result is assumed to be non-null.)